### PR TITLE
*: reject transfer leader when the region recently changed config

### PR DESCRIPTION
**Pull request by nolouch**
Signed-off-by: nolouch <nolouch@gmail.com>

## What have you changed? (mandatory)
Do not do transfer leader if the region recently has config change, which makes sure the region has time to apply pending config.

## What are the type of the changes? (mandatory)
record the recently changed region.

The currently defined types are listed below, please pick one of the types for this PR by removing the others:

- Improvement (change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
- Unit test

### CODE REVIEW

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -1316,14 +1317,15 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
                 self.fsm.peer.peer_heartbeats.insert(id, now);
                 if self.fsm.peer.is_leader() {
                     self.fsm.peer.peers_start_pending_time.push((id, now));
                 }
-                self.fsm.peer.recent_added_peer.update(id, now);
-                self.fsm.peer.insert_peer_cache(peer);
+                self.fsm
+                    .peer
+                    .recent_conf_change_region
+                    .update(cp.region.get_id(), now);
```
Comment (BusyJay) [inline src/raftstore/store/fsm/peer.rs]: Seems ID is unnecessary any more?

File: src/raftstore/store/peer.rs [hunk 2]
```diff
@@ -192,15 +192,15 @@ pub struct PeerStat {
     pub written_keys: u64,
 }
 
-pub struct RecentAddedPeer {
+pub struct RecentConfChangedPeer {
     pub reject_duration_as_secs: u64,
     pub id: u64,
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/peer.rs]: Id is no need any more.

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -1334,7 +1336,10 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
                         .peers_start_pending_time
                         .retain(|&(p, _)| p != peer_id);
                 }
-                self.fsm.peer.remove_peer_from_cache(peer_id);
+                self.fsm
```
Comment (siddontang) [inline src/raftstore/store/fsm/peer.rs]: em, seem we don't remove peer from the cache now?

**Comment by siddontang** [comment]
is it necessary to enhance the check that guaranteeing the transferee's commit index is larger than the index of ConfChange log?

@BusyJay

**Comment by BusyJay** [comment]
You can know what's followers' commit index currently.

File: tests/integrations/raftstore/test_conf_change.rs [hunk 4]
```diff
@@ -576,12 +576,26 @@ fn test_transfer_leader_safe<T: Simulator>(cluster: &mut Cluster<T>) {
...
-    cluster.reset_leader_of_region(region_id);
-    assert_ne!(cluster.leader_of_region(region_id).unwrap().get_id(), 3);
+    // transfer to all followers
+    let mut leader_id = cluster.leader_of_region(region_id).unwrap().get_id();
+    for peer in cluster.get_region(b"").get_peers() {
+        if peer.get_id() == leader_id {
+            continue;
+        }
+        cluster.transfer_leader(region_id, peer.clone());
+        cluster.reset_leader_of_region(region_id);
```
Comment (BusyJay) [inline tests/integrations/raftstore/test_conf_change.rs]: I don't think the test will fail without your changes.

Comment (nolouch) [reply tests/integrations/raftstore/test_conf_change.rs]: That ensures we every time the check the leader is not from the cache. I think it makes the tests not always be a success.

**Pull request merged**